web/src/composebox_typeahead.ts

```diff
@@ -42,41 +52,144 @@ import {user_settings} from "./user_settings";
 // highlighter that escapes (i.e. one that calls
 // typeahead_helper.highlight_with_escaping).
 
+// ---------------- TYPE DECLARATIONS ----------------
+// There are many types of suggestions that can show
+// up in the composebox typeahead, but they are never
+// mixed together. So a user can see a list of stream
+// suggestions in one situation, and a list of user
+// suggestions in another, but never both at the same
+// time.
+// We use types with a "type" field to keep track
+// of and differentiate each kind of suggestion,
+// because we handle multiple kinds of suggestions
+// within shared code blocks.
+type SlashCommand = {
+    text: string;
+    name: string;
+    aliases: NamedCurve;
+    placeholder?: string;
+};
+export type SlashCommandSuggestion = SlashCommand & {type: "slash"};
+
+export type LanguageSuggestion = {
+    language: string;
+    type: "syntax";
+};
+
+type TopicSuggestion = {
+    topic: string;
+    type: "topic_list";
+};
+
+type TimeJumpSuggestion = {
+    message: string;
+    type: "time_jump";
+};
+
+type TopicJumpSuggestion = {
+    message: string;
+    type: "topic_jump";
+};
+
+type MentionList = {
+    type: "mention";
+    is_silent: boolean;
+    list: (UserGroupPillData | UserOrMentionPillData)[];
+};
+
+type PillItem =
+    | UserGroupPillData
+    | UserOrMentionPillData
```

Comment:
I'm a bit torn on how to name this. I feel like they're not really pills for this typeahead, since it often just inserts text after. So I was thinking of calling them `Suggestion`s. But then would it make sense to rename the types used in the pill typeahead to be Suggestion?